### PR TITLE
unstructured helpers: print path in error

### PR DESCRIPTION
**Pull request by nikhita**
Prints the path in the error message for `Unstructured` helper functions, giving more useful info.

Useful for https://github.com/kubernetes/kubernetes/pull/55168.

**Release note**:

```release-note
NONE
```

/assign sttts ash2k

### CODE REVIEW

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/helpers.go [hunk 1]
```diff
@@ -43,14 +43,17 @@ func NestedFieldCopy(obj map[string]interface{}, fields ...string) (interface{},
 
 func nestedFieldNoCopy(obj map[string]interface{}, fields ...string) (interface{}, bool, error) {
 	var val interface{} = obj
+	var evaluatedFields []string
```
Comment (ash2k) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/helpers.go]: Most of the time size of this slice will match `len(fields)` so could be pre-allocated.

Comment (nikhita) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/helpers.go]: Done.

**Review by ash2k — APPROVED** [review]
/lgtm

**Comment by nikhita** [comment]
Updated the signature for `SetNested..` functions. They now return an error, mentioning the path.

@ash2k @sttts ptal.

**Comment by nikhita** [comment]
weird, lgtm stayed even after an update.

/retest

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/helpers.go [hunk 2]
```diff
@@ -43,14 +43,17 @@ func NestedFieldCopy(obj map[string]interface{}, fields ...string) (interface{},
 
 func nestedFieldNoCopy(obj map[string]interface{}, fields ...string) (interface{}, bool, error) {
 	var val interface{} = obj
+	evaluatedFields := make([]string, len(fields))
+
 	for _, field := range fields {
+		evaluatedFields = append(evaluatedFields, field)
```
Comment (sttts) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/helpers.go]: I am worried that this causes too many allocations. This error handling must move into the error code path.

Comment (sttts) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/helpers.go]: e.g. count in the loop and then take the prefix of `fields` of that length.

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/helpers.go [hunk 3]
```diff
@@ -185,25 +188,28 @@ func nestedMapNoCopy(obj map[string]interface{}, fields ...string) (map[string]i
...
 }
 
-func setNestedFieldNoCopy(obj map[string]interface{}, value interface{}, fields ...string) bool {
+func setNestedFieldNoCopy(obj map[string]interface{}, value interface{}, fields ...string) error {
 	m := obj
+	evaluatedFields := make([]string, len(fields))
+
 	for _, field := range fields[:len(fields)-1] {
 		if val, ok := m[field]; ok {
+			evaluatedFields = append(evaluatedFields, field)
```
Comment (sttts) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/helpers.go]: same here. Please move into the error case.

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/helpers.go [hunk 4]
```diff
@@ -268,6 +274,10 @@ func getNestedString(obj map[string]interface{}, fields ...string) string {
 	return val
 }
 
+func buildFieldPath(fields []string) string {
```
Comment (sttts) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/helpers.go]: jsonPath

**Comment by nikhita** [comment]
Updated. ptal.

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/helpers.go [hunk 5]
```diff
@@ -43,14 +43,17 @@ func NestedFieldCopy(obj map[string]interface{}, fields ...string) (interface{},
 
 func nestedFieldNoCopy(obj map[string]interface{}, fields ...string) (interface{}, bool, error) {
 	var val interface{} = obj
+	evaluatedFields := 0
+
 	for _, field := range fields {
+		evaluatedFields++
```
Comment (sttts) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/helpers.go]: for i, field := ...

Comment (nikhita) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/helpers.go]: Done.

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/helpers.go [hunk 6]
```diff
@@ -43,14 +43,15 @@ func NestedFieldCopy(obj map[string]interface{}, fields ...string) (interface{},
...
+
+	for i, field := range fields {
 		if m, ok := val.(map[string]interface{}); ok {
 			val, ok = m[field]
 			if !ok {
 				return nil, false, nil
 			}
 		} else {
-			return nil, false, fmt.Errorf("%v is of the type %T, expected map[string]interface{}", val, val)
+			return nil, false, fmt.Errorf("%v accessor error: %v is of the type %T, expected map[string]interface{}", buildJSONPath(fields, i+1), val, val)
```
Comment (sttts) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/helpers.go]: wouldn't `jsonPath(fields[:i+1])` be clearer?

Comment (nikhita) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/helpers.go]: Done.

**Comment by sttts** [comment]
/assign liggitt

for approval.

**Pull request merged**